pkg/infra/provider.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright Kurator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package infraprovider
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	awssdkcfn "github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	awsinfrav1 "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
+	awsbootstrapv1 "sigs.k8s.io/cluster-api-provider-aws/v2/cmd/clusterawsadm/api/bootstrap/v1beta1"
+	capabootstrap "sigs.k8s.io/cluster-api-provider-aws/v2/cmd/clusterawsadm/cloudformation/bootstrap"
+	cloudformation "sigs.k8s.io/cluster-api-provider-aws/v2/cmd/clusterawsadm/cloudformation/service"
+	"sigs.k8s.io/cluster-api-provider-aws/v2/util/system"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	infrav1 "kurator.dev/kurator/pkg/apis/cluster/v1alpha1"
+	"kurator.dev/kurator/pkg/infra/scope"
+	"kurator.dev/kurator/pkg/infra/template"
+	"kurator.dev/kurator/pkg/infra/util"
+	"kurator.dev/kurator/pkg/typemeta"
+)
+
+type Provider interface {
+	Reconcile(ctx context.Context, cluster *infrav1.Cluster) error
+	Clean(ctx context.Context, cluster *infrav1.Cluster) error
+}
+
+func NewProvider(kube client.Client, scope *scope.Cluster) Provider {
+	return &AWSProvider{
+		Kube:  kube,
+		scope: scope,
+	}
+}
+
+var _ Provider = &AWSProvider{}
+
+type AWSProvider struct {
```

Comment:
I do feel AWSProvider should be separated from the factory, and we may have other providers later

Reply:
yes, for now keep it here.